File: linear_model.py
#!/usr/bin/env python
# coding: utf-8

# This notebook demonstrates a simple linear regression analysis using [Python/R] to model Salary based on Years of Experience.
# 

# In[2]:


import pandas as pd
dataset = pd.read_csv("regression_data.csv")


# In[3]:


import matplotlib.pyplot as plt
plt.scatter(dataset["YearsExperience"], dataset["Salary"], color="red")


# In[4]:


from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(dataset[["YearsExperience"]], dataset[["Salary"]])


# In[5]:


plt.plot(dataset["YearsExperience"], model.predict(dataset[["YearsExperience"]]), color="blue")
plt.title("Salary vs Experience")
plt.xlabel("Years of Experience")
plt.ylabel("Salary")
plt.show()


# In[7]:


model.score(dataset[["YearsExperience"]], dataset[["Salary"]])  # R-squared


# **slope** and **intercept** of the fitted line 
# correlation coefficient

# In[ ]:


from scipy.stats import linregress
x = dataset["YearsExperience"]
y = dataset["Salary"]
slope, intercept, r_value, p_value, std_err = linregress(dataset["YearsExperience"], dataset["Salary"])
y_pred = slope * x + intercept


# **mse** between observed and predicted values

# In[ ]:





# In[7]:


from sklearn.metrics import mean_squared_error
mse = mean_squared_error(y, y_pred)


# In[8]:


print(slope)
print(intercept)
print(r_value)
print(mse)


# In[16]:


plt.plot(x, y_pred, label='Fitted line')
plt.text(0.05, 0.95, f"y = {slope:.2f}x + {intercept:.2f}\nr = {r_value:.2f}",
         transform=plt.gca().transAxes, fontsize=10, verticalalignment='top')
plt.legend()
plt.show()
plt.savefig("./linear_model.png")


# In[ ]:





# In[ ]:





# In[ ]:





# In[ ]:





# In[ ]:

File: linear_model.r
dataset <- read.csv("regression_data.csv")

plot(dataset$YearsExperience, dataset$Salary, col="red")

model <- lm(Salary ~ YearsExperience, data=dataset)

library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$YearsExperience, y = dataset$Salary), colour = 'red') +
  geom_line(aes(x = dataset$YearsExperience, y = predict(model, newdata = dataset)), colour = 'blue') +
  ggtitle('Salary vs Experience') +
  xlab('Years of experience') +
  ylab('Salary')

summary(model)

x <- dataset$YearsExperience
y <- dataset$Salary
model <- lm(y ~ x)
slope <- coef(model)[2]
intercept <- coef(model)[1]
r <- cor(x, y)

mse <- mean((predict(model) - y)^2)



print("intercept")
print(intercept)
print ("slope")
print (slope)
print("r")
print(r)
print("mse")
print(mse)

library(ggplot2)

p <-ggplot(dataset, aes(x = YearsExperience, y = Salary)) +
  geom_point() +
  geom_smooth(method = 'lm', se = FALSE) +
  annotate(
    "text",
    x = max(x) * 0.6,
    y = max(y) * 0.9,
    label = paste(
      "y =", round(slope, 2), "x +", round(intercept, 2),
      "\nr =", round(r, 2)
    ),
    hjust = 0
  ) +
  xlab("Years of Experience") +
  ylab("Salary")
ggsave("regression_plot_r.png",
       plot = p,
       width = 4, height = 2.5,
       units = "in", dpi = 150)






